platform/pulp/platform/models/content.py

```diff
@@ -0,0 +1,94 @@
+"""
+Content related Django models.
+"""
+import hashlib
+
+from django.contrib.contenttypes import fields
+from django.db import models
+
+
+from pulp.platform.models import Model, MasterModel, Notes
+from pulp.platform.models.storage import StoragePath
+
+
+class Content(MasterModel):
+    """
+    A piece of managed content.
+
+    :cvar natural_key_fields: Tuple of natural fields.  Must be: models.Field.
+    :type natural_key_fields: tuple
+
+    Fields:
+
+    :cvar type: The content type.
+    :type type: models.TextField
+
+    Relations:
+
+    :cvar notes: Arbitrary information stored with the content.
+    :type notes: fields.GenericRelation
+    """
+    natural_key_fields = ()
+
+    type = models.TextField(blank=False, default=None)
+
+    notes = fields.GenericRelation(Notes)
+
+    def natural_key(self):
+        return (getattr(self, f.name) for f in self.natural_key_fields)
+
+    def natural_key_digest(self):
+        h = hashlib.sha256()
+        for name in sorted(f.name for f in self.natural_key_fields):
+            h.update(name)
+            value = getattr(self, name)
+            if not isinstance(value, basestring):
+                h.update(str(value))
+            else:
+                h.update(value)
+        return h.hexdigest()
+
+
+class Artifact(Model):
```

Comment:
We'd need to talk through this but the `Artifact` and `LazyCatalogEntry` represent very different things.  The `Artifact` is a file related to a content unit.  The LazyCatalogEntry describes how an artifact can be downloaded.  *One* artifact can be referenced by *many* catalog entries (one for each importer that can download it).

The modeling of the artifact does mean we need less information in the catalog entry.  Each entry can be reduced to something like:
```
class LazyCatalogEntry(Model):
    url = models.TextField(blank=False, default=None)

    artifact = models.ForeignKey('Artifact', related_name='catalog', on_delete=models.CASCADE)
    importer = models.ForeignKey('RepositoryImporter', on_delete=models.CASCADE)
```

The downloading code can find the artifact based on `Artifact.file.name` which is the file storage path.  Then iterate though `artifact.catalog` just to get importers and urls that are used to download it.  As you mentioned, the checksum etc can just be gotten from the artifact.

The `FileField` supports `db_index=True` but not `unique` which may require a little work.